README.md
<!-- markdownlint-disable -->
<h1 align="center" style="margin:1em">
  <img src="./docs/static/logo.png"
       alt="Ghost in A Shell"
       width="200">
    <br/> Ghost in A shell
</h1>

<h4 align="center">
  Spawn your ghost blogs fearlessly
  <br /> (thank you Docker!)
</h4>

<p align="center">
  <a href="https://github.com/ebreton/ghost-in-a-shell/blob/master/CHANGELOG.md">
    <img src="https://img.shields.io/github/release/ebreton/ghost-in-a-shell.svg"
         alt="Changelog">
  </a>
  <a href="https://github.com/ebreton/ghost-in-a-shell/blob/master/LICENSE">
    <img src="https://img.shields.io/badge/license-MIT-blue.svg"
         alt="License" />
  </a>
</p>
<br>

ToC
--

<!-- TOC -->

- [Overview](#overview)
- [Installation and usage](#installation-and-usage)
- [Helpers for developers](#helpers-for-developers)
- [Production](#production)
- [Something is missing ?](#something-is-missing-)
- [Changelog](#changelog)
- [:warning: Backward incompatible changes](#warning-backward-incompatible-changes)
- [Contribution](#contribution)

<!-- /TOC -->

## Overview

You just want to try Ghost without messing up your computer? (and that's why you already have docker installed)

> simply run `make` and head to <http://localhost:3001>

You want to play a bit more, and you would like to have multiple Ghosts on your domain / server?

* Either use *make* again and again with different variables, to get your blogs on a **per-port basis**
    > e.g.: `NAME=another PORT=3002 make` for a second blog on <http://localhost:3002>
* or leverage the power of [traefik](https://traefik.io) and have multiple ghosts on a **per-path basis** (see section below)
    > e.g.: `NAME=yet-another make qa`. Head to <http://localhost/yet-another>

By the way, you would like to run an old version of Ghost ? `GHOST_VERSION=x.x.x make qa` will do this for you. You can use whatever version is available on [docker hub](https://hub.docker.com/_/ghost)

If you are worried with your data, be at rest: a local folder is created within path _./instances_, for every blog you create, named from $NAME variable. Stopping and restarting a blog with the same name will keep using the local data.

## Installation and usage

Command | Description | Remarks
---------|----------|---------
 `make` | straightforward (local) installation if you simply wish to bridge a container port on your host (3001 by default) | also known as `make dev`
 `make qa` | The former being not really convenient if you wish to serve on standard ports (80 or 443), and if you want anyone to access your blog easily, this command sets up a traefik router | recommended usage of [prod-stack](https://github.com/ebreton/prod-stack)
 <code>make&nbsp;prod</code> | Same as above, using a MariaDB instead of SQLite and an email provider (Mailgun) | requires a (free) account at Mailgun

You will find details and a step-by-step guide for both scenario in [INSTALL.md](./docs/INSTALL.md)

If you'd rather have a **Tutorial** than a raw INSTALL.md file, feel free to check my "_Ghost in A shell_" series:
1. [Part I : localhost](https://dev.to/ebreton/ghost-in-a-shell---part-i--localhost-5he9)
1. [Part II : HTTPs is the norm](https://dev.to/ebreton/ghost-in-a-shell---part-ii---https-is-the-norm-1jj4)
1. [Part III - To Production, with performance](https://dev.to/ebreton/ghost-in-a-shell---part-iii--to-production-with-performance-10g5)
2. [Part IV - Maintenance](https://dev.to/ebreton/ghost-in-a-shell---part-iv-and-final-maintenance-2a5p)

## Helpers for developers

Now that you have one (or more) blogs running, you might want to check their status, or access the containers...

A few helpers are provided within the Makefile:

Command | Description | Variables
---------|----------|---------
 `make vars` | Display the values of all env vars | All
 `make ps` | Display the running containers | None
 `make cli-version` | Display Ghost & ghost-cli version, + latest version available on docker hub | NAME
 `make shell` | Connect to given Ghost container | NAME
 `make logs` | Tail and follow the logs of given Ghost container | NAME
 `make stop` | Stop given Ghost container | NAME
 `make pull` | Update docker image from dockerhub | None
 `make restart` | Calls `make stop qa logs` | NAME (and from `make qa`: PROTOCOL, DOMAIN, URI)
 `make upgrade` | Calls `make pull restart` | same as above
 <code>make&nbsp;gatling&nbsp;&nbsp;&nbsp;&nbsp;</code> | Execute load tests | GATLING_BASE_URL, GATLING_USERS, GATLING_RAMP


More detailed can be found in [HELPERS.md](./docs/HELPERS.md)

## Production

Environment variables are defined in ./etc/prod.env, e.g:
- MYSQL_ROOT_PASSWORD
- MAILGUN_LOGIN
- MAILGUN_PASSWORD

Command | Description 
---------|----------
 <code>make&nbsp;prod&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</code> | Runs Ghost behind a proxy (Nginx+Traefik), on HTTPs, with a (Maria)DB and Mailgun email provider
 `make restart-prod` | Calls `make stop prod logs`
 `make upgrade-prod` | Calls `make pull restart-prod`


## Something is missing ?

Head to [github issues](https://github.com/ebreton/ghost-in-a-shell/issues) and submit one! Be sure to have a look at the [CONTRIBUTING.md](./docs/CONTRIBUTING.md) guide before


## Changelog

All notable changes to this project are documented in [CHANGELOG.md](./CHANGELOG.md).

## :warning: Backward incompatible changes

- :boom: v0.2.0 : data volumes are now created within subfolder ./instances. You will need to move your data from the root folder into this new place
- :alarm_clock: v0.4.0: `make traefik` replaced with `make qa` instead

## Contribution

Check out [CONTRIBUTING.md](./docs/CONTRIBUTING.md) for more details

As well as our [CODE_OF_CONDUCT.md](./docs/CODE_OF_CONDUCT.md), where we pledge to making participation in our project and our community a harassment-free experience for everyone


CHANGELOG.md
# Change Log

## [qa-release](https://github.com/ebreton/ghost-in-a-shell/tree/qa-release) (2018-06-13)
[Full Changelog](https://github.com/ebreton/ghost-in-a-shell/compare/0.3.0...qa-release)

## [0.3.0](https://github.com/ebreton/ghost-in-a-shell/tree/0.3.0) (2018-06-13)
[Full Changelog](https://github.com/ebreton/ghost-in-a-shell/compare/0.2.2...0.3.0)

**Merged pull requests:**

- All in one configurations [\#11](https://github.com/ebreton/ghost-in-a-shell/pull/11) ([ebreton](https://github.com/ebreton))

## [0.2.2](https://github.com/ebreton/ghost-in-a-shell/tree/0.2.2) (2018-06-04)
[Full Changelog](https://github.com/ebreton/ghost-in-a-shell/compare/0.2.1...0.2.2)

**Merged pull requests:**

- Support multiple versions of Ghost [\#10](https://github.com/ebreton/ghost-in-a-shell/pull/10) ([ebreton](https://github.com/ebreton))

## [0.2.1](https://github.com/ebreton/ghost-in-a-shell/tree/0.2.1) (2018-05-28)
[Full Changelog](https://github.com/ebreton/ghost-in-a-shell/compare/0.2.0...0.2.1)

**Merged pull requests:**

- Added make gatling for load tests [\#9](https://github.com/ebreton/ghost-in-a-shell/pull/9) ([ebreton](https://github.com/ebreton))
- added env variables for DB and Emails [\#8](https://github.com/ebreton/ghost-in-a-shell/pull/8) ([ebreton](https://github.com/ebreton))

## [0.2.0](https://github.com/ebreton/ghost-in-a-shell/tree/0.2.0) (2018-05-26)
[Full Changelog](https://github.com/ebreton/ghost-in-a-shell/compare/0.1.2...0.2.0)

**Merged pull requests:**

- Added make restart/upgrade and subfolder instances [\#7](https://github.com/ebreton/ghost-in-a-shell/pull/7) ([ebreton](https://github.com/ebreton))
- Added make pull [\#6](https://github.com/ebreton/ghost-in-a-shell/pull/6) ([ebreton](https://github.com/ebreton))

## [0.1.2](https://github.com/ebreton/ghost-in-a-shell/tree/0.1.2) (2018-05-23)
[Full Changelog](https://github.com/ebreton/ghost-in-a-shell/compare/0.1.1...0.1.2)

**Merged pull requests:**

- Added HELPERS.md to describe make commands [\#5](https://github.com/ebreton/ghost-in-a-shell/pull/5) ([ebreton](https://github.com/ebreton))
- Extended env vars [\#4](https://github.com/ebreton/ghost-in-a-shell/pull/4) ([ebreton](https://github.com/ebreton))
- More documentation [\#3](https://github.com/ebreton/ghost-in-a-shell/pull/3) ([ebreton](https://github.com/ebreton))

## [0.1.1](https://github.com/ebreton/ghost-in-a-shell/tree/0.1.1) (2018-05-22)
[Full Changelog](https://github.com/ebreton/ghost-in-a-shell/compare/0.1.0...0.1.1)

**Merged pull requests:**

- Enhanced management of environment vars [\#2](https://github.com/ebreton/ghost-in-a-shell/pull/2) ([ebreton](https://github.com/ebreton))
- Added release process [\#1](https://github.com/ebreton/ghost-in-a-shell/pull/1) ([ebreton](https://github.com/ebreton))

## [0.1.0](https://github.com/ebreton/ghost-in-a-shell/tree/0.1.0) (2018-05-22)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

etc/gatling-conf.scala
package ghost

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import scala.concurrent.duration._

class GhostFrontend extends Simulation {
  // Expected parameters: -Dusers=10 -Dramp=20 -DbaseUrl=...
  val nbUsers = Integer.getInteger("users", 10)
  val myRamp = java.lang.Long.getLong("ramp", 20)

  // Will request default posts created by ghost install, with 
  // 1- Base URL (Needs to be passed through JAVA_OPTS="-DbaseUrl=...")
  val baseUrl = System.getProperty("baseUrl")

  // 2- Posts URIs
  // https://stackoverflow.com/q/35730086
  val uriFeeder = Array(
    Map("URIKey" -> s"welcome"),
    Map("URIKey" -> s"the-editor"),
    Map("URIKey" -> s"using-tags"),
    Map("URIKey" -> s"managing-users"),
    Map("URIKey" -> s"private-sites"),
    Map("URIKey" -> s"advanced-markdown"),
    Map("URIKey" -> s"themes")
  ).random

  // HTTP connection setup
  val header  = "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
  val httpConf = http
    .baseURL(baseUrl)
    .inferHtmlResources()
    .acceptHeader(header)
    .doNotTrackHeader("1")
    .acceptLanguageHeader("en-US,en;q=0.5")
    .acceptEncodingHeader("gzip, deflate")
    .userAgentHeader("Gatling")
    .disableCaching
    .disableClientSharing
    .maxConnectionsPerHostLikeChrome

  // A scenario is a chain of requests and pauses
  val scn = scenario("GhostFrontend")
    .feed(uriFeeder)
    .exec(http("GET /${URIKey}")
    .get("/${URIKey}/")
  )

  // Execute
  setUp(
    scn.inject(
      atOnceUsers(nbUsers),
      constantUsersPerSec(5) during (myRamp seconds),
      rampUsers(nbUsers) over (myRamp seconds),
      heavisideUsers(10) over (myRamp seconds)
    ).protocols(httpConf)
  )
}


find_latest_versions.py
#!/usr/bin/python

from distutils.version import LooseVersion

import argparse
import logging
import requests
import re

from update_release import  set_logging_config
from versions import _version

session = requests.Session()

# authorization token
TOKEN_URL = "https://auth.docker.io/token?service=registry.docker.io&scope=repository:%s:pull"

# find all tags
TAGS_URL =  "https://index.docker.io/v2/%s/tags/list"
TAG_RE = re.compile("^[\d]+(\.[\d]+)*$")

# get image digest for target
TARGET_DIGEST = "https://index.docker.io/v2/%(repository)s/manifests/%(tag)s"

class Fetcher:

    DIGEST_HEADER = {}

    def __init__(self, repository):
        self.repository = repository
        self.token = self.get_token()
        self.headers = {"Authorization": "Bearer %s"% self.token}
        self.headers_for_tags = {
            "Authorization": "Bearer %s"% self.token,
            "Accept": "application/vnd.docker.distribution.manifest.v2+json"
        }
        logging.debug("initialized fetcher for %s", self.repository)


    def get_token(self):
        response = session.get(TOKEN_URL % self.repository)
        response.raise_for_status()
        token = response.json().get("token")
        logging.debug("got token: %s", token)
        return token


    def get_versions(self):
        response = session.get(TAGS_URL % self.repository, headers=self.headers_for_tags)
        response.raise_for_status()
        all_tags = response.json().get("tags")
        numbered_tags = filter(lambda x: TAG_RE.match(x), all_tags)
        versions = map(LooseVersion, numbered_tags)
        logging.debug("got tags: %s", versions)
        return versions


def find_latest(repository):
    fetcher = Fetcher(repository)
    all_tags = fetcher.get_versions()
    return max(all_tags)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        usage="""Version checker script

This file retreives the latest version of ghost container image from docker hub
It can be run with both python 2.7 and 3.6""")
    parser.add_argument("repository", nargs='?', 
        help="repository name [default:library/ghost]",
        default="library/ghost")
    parser.add_argument('-v', '--version', action='store_true')
    parser.add_argument('-d', '--debug', action='store_true')
    parser.add_argument('-q', '--quiet', action='store_true')
    args = parser.parse_args()

    set_logging_config(quiet=args.quiet, debug=args.debug)
    logging.debug(args)

    # version needs to be print to output in order to be retrieved by Makefile
    if args.version:
        print(_version)
        raise SystemExit()

    print(find_latest(args.repository))


update_release.py
#!/usr/bin/python
"""Release manager script

This file should be used as post-commit in .git/hooks
It can be run with both python 2.7 and 3.6

Usage:
    commands.py [-q | -d]
    commands.py confirm [-q | -d]
    commands.py publish [-q | -d]
    commands.py check [--branch=BRANCH] [-q | -d]
    commands.py -h
    commands.py -v

Options:
    -h, --help       display this message and exit
    -v, --version    display version
    --branch=BRANCH  branch name to check for (default to master)
    -q, --quiet      set log level to WARNING (instead of INFO)
    -d, --debug      set log level to DEBUG (instead of INFO)
"""
import os
import json
import logging
import subprocess
import sys
import shutil
import argparse
from pprint import pprint

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
COPY_PATH = os.path.sep.join([BASE_DIR, 'src', 'myapp', 'versions.py'])

if BASE_DIR.endswith('hooks'):
    BASE_DIR = os.path.abspath(os.path.sep.join(
        [BASE_DIR, '..', '..']))
    sys.path.append(BASE_DIR)

from versions import __file__ as RELEASE_FILE  # noqa
from versions import _release, _build, _version  # noqa


def set_logging_config(quiet=False, debug=False):
    """
    Set logging with the 'good' level

    Arguments keywords:
    kwargs -- list containing parameters passed to script
    """
    # set up level of logging
    level = logging.INFO
    if quiet:
        level = logging.WARNING
    elif debug:
        level = logging.DEBUG

    # set up logging to console
    logging.basicConfig(format='%(levelname)s - %(funcName)s - %(message)s')
    logger = logging.getLogger()
    logger.setLevel(level)


def compute():
    logging.debug("Updating versions...")

    # compute version from release
    try:
        # get last release & version
        release_version = list(map(int, _release.split('-')[0].split('.')))
        candidate_version = list(map(int, _version.split('-')[0].split('.')))
        # nothing to do if candidate_version is already more than release
        if candidate_version > release_version:
            logging.debug("version already set to %s. not changing it", candidate_version)
            version = _version
        # increment and convert back in strings, with suffic -rc
        else:
            logging.debug("incrementint version from release %s", release_version)
            major, minor, patch = release_version
            version = '.'.join(map(str, [major, minor, patch+1])) + "-rc"
    except Exception as err:
        logging.error("Exception occured while trying to generate new version: %s", err)
        version = _version

    logging.info("will set _version=%s", version)

    # get build and release numbers from git
    try:
        build = subprocess.check_output(
            ["git", "rev-parse", "HEAD"], cwd=BASE_DIR).decode('utf-8').strip()

        release = subprocess.check_output(
            ["git", "describe", "--tags", "--match", "[0-9]*"], cwd=BASE_DIR).decode('utf-8').strip()
    except Exception as err:
        logging.warning("Using previous build & release, since git does not seem available: %s", err)
        release = _release
        build = _build

    logging.info("will set _build=%s", build)
    logging.info("will set _release=%s", release)

    # update file
    with open(RELEASE_FILE, 'w') as output:
        content = """# flake8: noqa
# This file is autognerated by post-commit hook

# the release comes from git and should not be modified
# => read-only
_release = '{0}'

# you can set the next version number manually
# if you do not, the system will make sure that version > release
# => read-write, >_release
_version = '{1}'

# the build number will generate conflicts on each PR merge
# just keep yours every time
# => read-only
_build = '{2}'""".format(release, version, build)
        output.write(content)


try: input = raw_input
except: pass


def confirm_release():
    print("version currently set to {}".format(_version))
    answer = ""
    while answer not in ["y", "n"]:
        answer = input("OK to push to continue [Y/N]? ").lower()
    if answer != "y":
        raise SystemExit("Please confirm version number to continue")


def confirm_push():
    try:
        branch = subprocess.check_output(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"], cwd=BASE_DIR).decode('utf-8').strip()
        print("\nYou are on branch '{}'".format(branch))
        print("on version        '{}'".format(_version))
        print("\nType [y] to comfirm push".format(branch))
        answer = input("or any other key to abort:").lower()
        if answer != "y":
            raise SystemExit("Push aborted...")
    except Exception as err:
        logging.warning("Git does not seem available: %s", err)
        raise SystemExit("This command requires git")


def publish(dry_run=False):
    """ POST /repos/:owner/:repo/releases
        
        https://developer.github.com/v3/repos/releases/#create-a-release
    """
    # dynamic import to allow the other commands to run without requests
    import requests

    # get gihub config. If not set -> POST will fail, developer will understand
    github_owner = os.environ.get('GITHUB_OWNER')
    github_repo = os.environ.get('GITHUB_REPO')
    github_user = os.environ.get('GITHUB_USER')
    github_key = os.environ.get('GITHUB_KEY')

    # build request
    url = "https://api.github.com/repos/{}/{}/releases".format(github_owner, github_repo)
    changelog_url = "https://github.com/{}/{}/blob/release-{}/CHANGELOG.md".format(github_owner, github_repo, _version)
    post_args = {
        "tag_name": _version,
        "name": "Release {}".format(_version),
        "body": "See [CHANGELOG.md]({}) for all details".format(changelog_url),
        "draft": False,
        "prerelease": False
        }
    logging.debug("POST %s with data: %s", url, post_args)

    # make request and raise exception if we had an issue
    if not dry_run:
        response = requests.post(url, data=json.dumps(post_args), auth=(github_user, github_key))
        response.raise_for_status()
    else:
        print("POST {}".format(url))
        print("auth({}, xxx)".format(github_user))
        pprint(post_args)



def check_branch(expected='master'):
    try:
        current = subprocess.check_output(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"], cwd=BASE_DIR).decode('utf-8').strip()
        if current != expected:
            raise SystemExit("You are in {}, whereas expected branch is {}".format(current, expected))
        logging.info("You are in {}".format(current))
    except Exception as err:
        logging.warning("Git does not seem available: %s", err)
        raise SystemExit("This command requires git")


if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        usage="""Release manager script

This file should be used as post-commit in .git/hooks
It can be run with both python 2.7 and 3.6""")
    parser.add_argument("command", nargs='?', help="[confirm|publish|check]")
    parser.add_argument('--prod', action='store_true', help="used with command confirm")
    parser.add_argument('--dry-run', action='store_true', help="used with command publish")
    parser.add_argument('--branch', help="used with command check_branch")
    parser.add_argument('-v', '--version', action='store_true')
    parser.add_argument('-d', '--debug', action='store_true')
    parser.add_argument('-q', '--quiet', action='store_true')
    args = parser.parse_args()

    set_logging_config(quiet=args.quiet, debug=args.debug)
    logging.debug(args)

    # version needs to be print to output in order to be retrieved by Makefile
    if args.version:
        print(_version)
        raise SystemExit()

    if args.command == 'confirm':
        if args.prod:
            confirm_push()
        else:
            confirm_release()
    elif args.command == 'check':
        check_branch(expected=args.branch)
    elif args.command == 'publish':
        publish(dry_run=args.dry_run)
    else:
        compute()
